salesforce/metadataApi.go

```diff
@@ -0,0 +1,268 @@
+package salesforce
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/amp-labs/connectors/common"
+	"golang.org/x/oauth2"
+)
+
+const (
+	versionPrefix             = "v"
+	openParenthesis           = "<"
+	closeParenthesis          = ">"
+	closeWithSlashParenthesis = "/>"
+)
+
+var ErrCreateMetadata = fmt.Errorf("error in CreateMetadata")
+
+type XMLSchema interface {
+	ToXML() string
+}
+
+type XMLAttributes struct {
+	Key   string `json:"key"`
+	Value string `json:"value"`
+}
+
+func (x *XMLAttributes) ToXML() string {
+	return fmt.Sprintf(`%s="%s"`, x.Key, x.Value)
+}
+
+type XMLString string
+
+func (x XMLString) ToXML() string {
+	return string(x)
+}
+
+type XMLData struct {
+	XMLName     string           `json:"xmlName"`
+	Attributes  []*XMLAttributes `json:"attributes"`
+	Children    []XMLSchema      `json:"children"`
+	SelfClosing bool             `json:"selfClosing"`
+}
+
+//nolint:cyclop
+func (x *XMLData) UnmarshalJSON(b []byte) error {
+	data := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(b, &data); err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(data["xmlName"], &x.XMLName); err != nil {
+		return err
+	}
+
+	if err := json.Unmarshal(data["attributes"], &x.Attributes); err != nil {
+		return err
+	}
+
+	children := []interface{}{}
+	if err := json.Unmarshal(data["children"], &children); err != nil {
+		return err
+	}
+
+	for _, child := range children {
+		if childValue, ok := child.(string); ok {
+			x.Children = append(x.Children, XMLString(childValue))
+
+			continue
+		}
+
+		if childValue, ok := child.(map[string]interface{}); ok {
+			childData, err := json.Marshal(childValue)
+			if err != nil {
+				return err
+			}
+
+			childXML := &XMLData{}
+			if err := json.Unmarshal(childData, childXML); err != nil {
+				return err
+			}
+
+			x.Children = append(x.Children, childXML)
+
+			continue
+		}
+	}
+
+	if err := json.Unmarshal(data["selfClosing"], &x.SelfClosing); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (x *XMLData) ToXML() string {
+	start := x.startTag()
+	if x.SelfClosing {
+		return start
+	}
+
+	end := x.endTag()
+
+	chilren := []string{}
+	for _, child := range x.Children {
+		chilren = append(chilren, child.ToXML())
+	}
+
+	return fmt.Sprintf("%s%s%s", start, strings.Join(chilren, ""), end)
+}
+
+func (x *XMLData) startTag() string {
```

Comment:
Can we use [encoding/xml](https://gobyexample.com/xml) to implement this? It should provide easy & tested translation between json & xml.

Reply:
I have done researches on that package, but the pacakge has many limitations that I found we could not use it for our use cases. With that package  I ran into problem was that we have to define struct for every object exists in Saleforce, plus there is no clear way to create custom object since we can't define struct for every custom object. 

My approach is to make the xml data structure to be more of wrapper container that can be used for any sort of object regardless of type. The way we would get definition for custom objects will be in yaml file, which need to be translated to some json. 

This data structure I created can be easily transformed from json, or even from yaml we can make it to transform easier since the underliying structure has json representation.